Fixes #24040 - Move the search box to React

Ron-Lavi opened the pull request:
You are welcome to check the component's **[Storybook](https://laviro.github.io/foreman/?selectedKind=Components%2FSearchBar&selectedStory=Search%20Bar&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Factions%2Factions-panel)**

### **Usage example along with Redux lifecycle**
 
![searchbar](https://user-images.githubusercontent.com/26363699/44529135-f9063f00-a6f3-11e8-879c-6aca0ab1fc32.gif)

ronlavi2412 committed:
Fixes #24040 - Move the search box to React

ohadlevy commented on app/views/common/_searchbar.html.erb:
```diff
@@ -3,19 +3,30 @@
+  <div id="auto-complete"></div>
 <% end %>
 
 <%= mount_react_component('BookmarkContainer', '#bookmark-form', {
```
I really think there should be only one component, that includes both search and bookmarks.

ohadlevy commented on package.json:
```diff
@@ -6,7 +6,6 @@
   "private": true,
   "devDependencies": {
     "@storybook/addon-actions": "^3.2.12",
-    "@storybook/addon-knobs": "^3.4.3",
```
why?

ohadlevy commented on package.json:
```diff
@@ -22,7 +21,7 @@
     "babel-preset-react": "^6.5.0",
     "babel-register": "^6.9.0",
-    "compression-webpack-plugin": "~1.1.11",
+    "compression-webpack-plugin": "~0.3.1",
```
why?

ohadlevy commented on package.json:
```diff
@@ -74,10 +73,11 @@
     "multiselect": "~0.9.12",
     "patternfly": "^3.42.0",
-    "patternfly-react": "2.5.1",
+    "patternfly-react": "^2.10.2",
```
please pin to a specific version.

ohadlevy commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoCompleteActions.js:
```diff
@@ -0,0 +1,20 @@
+export const getOptions = query => (dispatch) => {
+  const q = query.replace(/\s/g, '+').replace(/=/g, '%3D');
+
+  const url = `models/auto_complete_search?search=${q}`;
```
please pass the url prefix (e.g. models) as an argument instead of hard code it.

ohadlevy commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoCompleteActions.js:
```diff
@@ -0,0 +1,20 @@
+    successAction: AUTO_COMPLETION_SUCCESS,
+    failedAction: AUTO_COMPLETION_FAILURE,
+    url,
+    item: { url, query },
```
probably useful to keep the screen name (or object) name too, so it would show up in the redux actions

ohadlevy commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoCompleteReducer.js:
```diff
@@ -0,0 +1,41 @@
+  options: [],
+  status: 'PENDING',
+  error: null,
+  searchQuery: '',
```
do we care about the original search query vs the current search query? (assuming one starts typing and change the existing filter prior to clicking on search)

ohadlevy commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoCompleteReducer.js:
```diff
@@ -0,0 +1,41 @@
+    case AUTO_COMPLETION_REQUEST:
+      return state
+        .set('error', null)
+        .set('status', STATUS.PENDING);
```
please use merge instead of multiple sets

ohadlevy commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoCompleteSearchBox.js:
```diff
@@ -0,0 +1,117 @@
+import React from 'react';
+import _ from 'lodash';
+import { TypeAheadSelect } from 'patternfly-react';
+import { Menu, MenuItem, Highlighter } from 'react-bootstrap-typeahead';
```
should this be a PR to patternfly (so its exposed?)

ohadlevy commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoCompleteSearchBox.js:
```diff
@@ -0,0 +1,117 @@
+                    onClick={this.handleClear}
+                    title=""
+                    data-original-title="Clear">
+                    ×
```
please use ```&times;``` instead.

amirfefer replied:
I've used [icon](https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/common/SearchInput/index.js#L35)

ohadlevy commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/auto-complete.scss:
```diff
@@ -0,0 +1,99 @@
+$color-pf-black-100:             #fafafa;
```
are these not defined already in patternfly?

ohadlevy commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/index.js:
```diff
@@ -0,0 +1,16 @@
+import AutoCompleteSearchBox from './AutoCompleteSearchBox';
+import reducer from './AutoCompleteReducer';
+
+
```
nitpick: extra line

ohadlevy commented on webpack/assets/javascripts/react_app/components/SearchButton/SearchButton.js:
```diff
@@ -0,0 +1,19 @@
+    <Button
+        onClick={() => handleClick(searchQuery)}>
+        <Icon name="search" />
+        {' Search'}
```
why the extra space?

amirfefer replied:
translation is missing

ohadlevy commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoCompleteSearchBox.js:
```diff
@@ -0,0 +1,117 @@
+    this.handleItemSelect = this.handleItemSelect.bind(this);
+    this.handleInputFocus = this.handleInputFocus.bind(this);
+    this.handleClear = this.handleClear.bind(this);
+    this.renderMenu = this.renderMenu.bind(this);
```
can we use this [style](https://github.com/theforeman/foreman/pull/5771/files#diff-315e6a5832a6d4bbc35c97b46b953729R13)  instead?

ohadlevy commented on webpack/assets/javascripts/react_app/components/SearchButton/SearchButton.js:
```diff
@@ -0,0 +1,19 @@
+const handleClick = (searchQuery) => {
+  const uri = new URI(window.location.href);
+  uri.setSearch('search', searchQuery.trim());
+  window.Turbolinks.visit(uri.toString());
```
don't we have duplication here with https://github.com/theforeman/foreman/pull/5771/files#diff-ccc2dd1b48def42815b6974e5a635e4a? and tfm.tools...

ohadlevy reviewed (commented):
initial code only review - thanks!

amirfefer commented on app/views/common/_searchbar.html.erb:
```diff
@@ -3,19 +3,30 @@
   <div class="input-group">
+  <% if controller_name == "models" %>
+    <span id="auto-complete"></span>
+    <%= mount_react_component('AutoCompleteSearch', '#auto-complete', {:search => request.query_parameters['search']}.to_json) %>
```
IMHO this is too specific, what would we do when we'd like to add another page ?
would it be better to mount it in the hw model index view? (and override the older auto_complete)
afaik you can use `params[]` instead of `request.query_parameters['']`

amirfefer commented on app/views/common/_searchbar.html.erb:
```diff
@@ -3,19 +3,30 @@
     <span class="input-group-btn">
+    <% if controller_name == "models" %>
+      <span id="search-button"></span>
+      <%= mount_react_component('SearchButton', '#search-button', {}.to_json) %>
```
ditto

amirfefer commented on app/views/common/_searchbar.html.erb:
```diff
@@ -3,19 +3,30 @@
       <span class="dropdown" id="bookmark-form"></span>
     </span>
   </div>
+  <div id="auto-complete"></div>
```
this id already used (line 7)

amirfefer commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoCompleteActions.js:
```diff
@@ -0,0 +1,20 @@
+import { ajaxRequestAction } from '../../redux/actions/common';
+
+export const getOptions = query => (dispatch) => {
+  const q = query.replace(/\s/g, '+').replace(/=/g, '%3D');
```
could you use informative variables names?

amirfefer commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoCompleteSearchBox.js:
```diff
@@ -0,0 +1,117 @@
+import React from 'react';
+import _ from 'lodash';
```
could you import the specific functions from lodash?

amirfefer commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoCompleteReducer.js:
```diff
@@ -0,0 +1,41 @@
+      return state
+        .set('options', Object.keys(payload)
+          .filter(k => !['query', 'url'].includes(k))
+          .map(k => payload[k]))
```
IMHO it would much better to use selectors with memorization hete (reselect?), this probably will invoked many times.

amirfefer replied:
I've added `reselect` in [this PR](https://github.com/theforeman/foreman/pull/5778)

amirfefer reviewed (changes requested):
@Laviro Thanks! I left some inline comments

ohadlevy commented:
@Laviro thanks! I've started some refactoring at https://github.com/ohadlevy/foreman/tree/5771 - these are suggestions i saw as i started looking though the code...

Ron-Lavi commented:
Thanks @ohadlevy & @amirfefer for your very detailed reviews !!

I have fixed most of them but still have few to work on.
I hope that my last commit did answer some of your questions, 
promise to answer the rest of them soon :)

tstrachota commented:
Hi @Laviro, I didn't read or check your code closely so please pardon my ignorance:
How is it different from Katello's [TypeAhead](https://github.com/Katello/katello/tree/master/webpack/move_to_pf/TypeAhead) and [Search](https://github.com/Katello/katello/tree/master/webpack/components/Search)?
Can their code be reused?
Can we make sure that Katello will be able to consume this new solution?

boaz0 commented:
@tstrachota 
IIRC, this patch is using [`patternfly-react`'s TypeAhead](https://rawgit.com/patternfly/patternfly-react/gh-pages/index.html?knob-Submit%20Form%20On%20Enter=true&knob-select%20Hint%20On%20Enter=true&knob-Set%20the%20Minimum%20Length%20to%20Autocomplete=1&knob-highlight%20Only%20Result=true&knob-Allow%20Adding%20New%20Items=true&knob-Multiple%20Selection=true&knob-Clear%20Button=true&selectedKind=patternfly-react%2FForms%20and%20Controls%2FTypeAhead%20Select&selectedStory=Highlight%20Results&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs).

Because `patternfly-react` typeahead is built on top of [`react-bootstrap-typeahead`](https://github.com/ericgio/react-bootstrap-typeahead) there is a big diff between the two implementations.

amirfefer commented on app/views/common/_searchbar.html.erb:
```diff
@@ -1,21 +1,18 @@
-  documentationUrl: documentation_url("4.1.5Searching")
-}.to_json) %>
+<%= mount_react_component('SearchBar', '.search-bar', {
+    form: {
```
extract this to a helper?

amirfefer commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoCompleteActions.js:
```diff
@@ -0,0 +1,20 @@
+import {
+  AUTO_COMPLETION_REQUEST,
```
`AUTO_COMPLETE_REQEST` would be better IMO

amirfefer commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoCompleteActions.test.js:
```diff
@@ -0,0 +1,34 @@
+
+describe('AutoComplete actions', () => {
+  it('getOptions should call ajaxRequestAction with search query', () => {
+    const query = 'name = model2';
```
extract this to a fixtures file?

amirfefer commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoCompleteReducer.js:
```diff
@@ -0,0 +1,38 @@
+
+export default (state = initialState, action) => {
+  const { payload, type } = action;
+  const optionsSelector = createSelector(
```
selectors should be on different files (i.e `SelecorAutoComplete.js`)
and should be used through `mapStateToProps `

You can follow this PR #5778

amirfefer commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoCompleteReducer.js:
```diff
@@ -0,0 +1,38 @@
+      return Immutable.merge(state, {
+        error: null,
+        status: STATUS.RESOLVED,
+        options: optionsSelector(),
```
afaik this should be in `mapStateToProps`

amirfefer commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoCompleteReducer.js:
```diff
@@ -0,0 +1,38 @@
+        searchQuery: payload.query,
+      });
+    case AUTO_COMPLETION_FAILURE:
+      return Immutable.merge(state, { error: payload.error, status: STATUS.ERROR });
```
I think you can call merge on state:  `state.merge(...)`

amirfefer commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoCompleteReducer.test.js:
```diff
@@ -0,0 +1,63 @@
+
+const getMockOptions = () => ([
+  {
+    completed: '', part: ' hardware_model ', label: ' hardware_model ', category: '',
```
extract this to a fixutres file?

amirfefer replied:
?

Ron-Lavi replied:
I will start soon to deal with tests :)

amirfefer commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoCompleteSearchBox.js:
```diff
@@ -0,0 +1,77 @@
+class AutoCompleteSearchBox extends React.Component {
+  constructor(props) {
+    super(props);
+    this.handleInputChange = debounce(this.handleInputChange, 250);
```
can you use `DebounceInput` from `react-debounce-input`
 you can see its usage at `SearchInput` componen

Ron-Lavi replied:
I don't think this will work in our case because the `react-bootstrap-typeahead` 
events are raised by inner-state updates nor window events.. 
Another problem is that we can't wrap or replace the inner input field with `DebounceInput`.

amirfefer commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/helpers.js:
```diff
@@ -0,0 +1,41 @@
+  return map;
+};
+
+export const renderMenu = (results, menuProps) => {
```
This entire filtering would invoked each time the input render.
you already use selectors, is there an option to move some filtering from here ?

amirfefer commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/helpers.js:
```diff
@@ -0,0 +1,41 @@
+      category[1].map((option) => {
+        const item =
+          <MenuItem key={idx} option={option.label} position={idx}>
+            <Highlighter search={menuProps.text}>
```
There is a component that probably does the same:
https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/common/SubstringWrapper/index.js#L4
we can use `Highlighter` instead in other places

amirfefer commented on webpack/assets/javascripts/react_app/components/SearchBar/index.js:
```diff
@@ -0,0 +1,16 @@
+      </form>
+);
+
+export default SearchBar;
```
`PropTypes` is missing in some components

amirfefer commented on webpack/assets/javascripts/react_app/components/SearchButton/SearchButton.js:
```diff
@@ -0,0 +1,19 @@
+import React from 'react';
+import { Button, Icon } from 'patternfly-react';
+
+const handleClick = (searchQuery) => {
```
would it be better to handle this  in the component's consumer ? with this approach the `onClick` will be dynamic.

amirfefer commented on webpack/assets/javascripts/react_app/components/SearchButton/index.js:
```diff
@@ -0,0 +1,6 @@
+
+const mapStateToProps = ({ autocomplete }) => ({ searchQuery: autocomplete.searchQuery });
+
+export default connect(mapStateToProps)(SearchButton);
```
why connect this small component to redux? couldn't we  pass the `searchQuery` from the parent?

amirfefer reviewed (changes requested):
Looks great! @Laviro  :)
I left few inline comments

ohadlevy commented on app/views/common/_searchbar.html.erb:
```diff
@@ -1,21 +1,15 @@
+      searchQuery: request.query_parameters['search']
+    },
+    bookmarks: { 
+      controller: auto_complete_controller_name,
```
this is the same option for both autocomplete and bookmarks, maybe send it only once?

Ron-Lavi replied:
Nice ! I will pass only one controller :+1:

ohadlevy commented:
[ok to test]

ohadlevy commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoComplete.js:
```diff
@@ -0,0 +1,98 @@
+    super(props);
+    props.getInitialResults(props.data);
+    this.handleInputChange = debounce(this.handleInputChange, 250);
+    this.handleInputFocus = debounce(this.handleInputFocus, 250);
```
wouldn't initiate focus and mount will both generate api calls? TBH I would drop the initial results and just fetch the data on focus (similar to today).

Ron-Lavi replied:
I think the problem with focus is that it would fetch the same data many times,
and what we really need is getting the basic results once the page is loaded, and than update the results only on input changes.

ohadlevy replied:
so just query it once, if it was in focus and its in the right state make the api call?

ohadlevy commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoComplete.js:
```diff
@@ -0,0 +1,98 @@
+    props.getInitialResults(props.data);
+    this.handleInputChange = debounce(this.handleInputChange, 250);
+    this.handleInputFocus = debounce(this.handleInputFocus, 250);
+    this.handleLoading = debounce(this.handleLoading, 250);
```
since this is repeating it self, maybe wrap it in on function? something like https://github.com/ohadlevy/foreman/commit/5ca2081ee7896cfecb7a3a23ca54a4b8298a3f68#diff-1c7267c5369cb03d34ff470c6a0ece96R17

Ron-Lavi replied:
Great idea :+1: I wil add it to `common/helpers`

ohadlevy commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoComplete.js:
```diff
@@ -0,0 +1,98 @@
+    this.handleInputFocus = debounce(this.handleInputFocus, 250);
+    this.handleLoading = debounce(this.handleLoading, 250);
+
+    helpers.bindMethods(this, ['handleClear', 'handleInputChange', 'handleItemSelect', 'handleInputFocus', 'getResults']);
```
I personally prefer more es6 style of writing this kind of components without binding this, @sharvit ?

Ron-Lavi replied:
For some reason the linter raised errors when I tried the es6 style,
I will try again though :)

ohadlevy commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoComplete.js:
```diff
@@ -0,0 +1,98 @@
+          name="close"
+          onClick={this.handleClear}
+          title=""
+          data-original-title="Clear" />
```
String not extract, also, imho this is legacy pop over, if you want a popover, you should use the [popover/tooltip](https://rawgit.com/patternfly/patternfly-react/gh-pages/index.html?knob-Popover%20Title=Popover&knob-Popover%20Content=%3Cstrong%3EHoly%20guacamole%21%3C%2Fstrong%3E%20Check%20this%20info.&knob-Placement=right&knob-Trigger=hover%20focus&knob-Tooltip=%3Cstrong%3EHoly%20guacamole%21%3C%2Fstrong%3E%20Check%20this%20info.&selectedKind=patternfly-react%2FWidgets%2FTooltip&selectedStory=Tooltip&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs) component?

ohadlevy commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoCompleteActions.js:
```diff
@@ -0,0 +1,46 @@
+  AUTO_COMPLETE_REQUEST,
+  AUTO_COMPLETE_SUCCESS,
+  AUTO_COMPLETE_FAILURE,
+  AUTO_COMPLETE_INITIAL_RESULTS,
```
whats the difference between initial results and success ? can't they be combine into one?

Ron-Lavi replied:
You are absolutley right :+1:

ohadlevy commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoCompleteActions.js:
```diff
@@ -0,0 +1,46 @@
+};
+
+export const getInitialResults = ({ searchQuery, controller }) => (dispatch) => {
+  const url = `${controller}/auto_complete_search?search=${searchQuery}`;
```
This looks very similar to the getResults.. IMHO you should drop this one.

ohadlevy commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoCompleteActions.js:
```diff
@@ -0,0 +1,46 @@
+} from './AutoCompleteConstants';
+
+export const getResults = (query, controller, trigger = '') => (dispatch) => {
+  let url = `${controller}/auto_complete_search?search=`;
```
consider receiving the auto completer url as a prop from rails.

boaz0 replied:
Sounds good but when we move this page to React, will we still be able to receive it as props from rails?

ohadlevy replied:
yes, why not?

boaz0 replied:
From what I've heard we won't `mount` the component in `_searchbar.html.erb`.

ohadlevy replied:
you will mount it from some where, in either way, its not a bad idea to have it defined only once.

boaz0 replied:
I am not arguing about that just wanted to know more details. if we can do it, then "lets do it!" :grinning:

ohadlevy commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoCompleteActions.js:
```diff
@@ -0,0 +1,46 @@
+  if (query) {
+    searchQuery = clearSpaces(query);
+    const uriQuery = searchQuery.replace(/\s/g, '+').replace(/=/g, '%3D');
+    url += `${uriQuery}+`;
```
consider using urijs (as used in other parts of the codebase)

ohadlevy commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/helpers.js:
```diff
@@ -0,0 +1,45 @@
+import React from 'react';
+import { Menu, MenuItem } from 'react-bootstrap-typeahead';
```
is there another way to get these without relaying on the typeahead package? should we expose it on PF instead?

boaz0 replied:
Yep, we should expose it from PF.

Ron-Lavi replied:
Going to push a [PR](https://github.com/patternfly/patternfly-react/pull/497) to PF on this issue

ohadlevy commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/helpers.js:
```diff
@@ -0,0 +1,45 @@
+import { Menu, MenuItem } from 'react-bootstrap-typeahead';
+import SubstringWrapper from '../common/SubstringWrapper';
+
+export const groupBy = (list, keyGetter) => {
```
is this is the same as lodash groupby?

boaz0 replied:
Yes, it does the same thing. @Laviro use [`lodash`'s `groupBy`](https://lodash.com/docs/#groupBy)
change line 24 to
```js
const grouped = _.groupBy(list, r => r.category)
```

ohadlevy replied:
or rather, avoid using ```_``` and import it directly just like in https://github.com/theforeman/foreman/pull/5771/files#diff-58cdf6e5ce7e5afcd2cd978a1bb22067R3

ohadlevy commented on webpack/assets/javascripts/react_app/components/componentRegistry.js:
```diff
@@ -68,6 +69,7 @@ const coreComponets = [
   { name: 'PasswordStrength', type: PasswordStrength },
   { name: 'BreadcrumbBar', type: BreadcrumbBar },
   { name: 'FactChart', type: FactChart },
+  { name: 'SearchBar', type: SearchBar },
```
please remove the bookmarks component from the mounter.

ohadlevy commented on webpack/assets/javascripts/react_app/components/SearchBar/SearchBar.js:
```diff
@@ -0,0 +1,37 @@
+const handleSearch = (searchQuery = '') => {
+  const uri = new URI(window.location.href);
+  uri.setSearch('search', searchQuery.trim());
+  window.Turbolinks.visit(uri.toString());
```
@ripcurld should we reuse updateTable ? (just wondering about the table /pagination usage cases)

boaz0 replied:
1) We can use `updateTable` here.
2) However, when `Table` or `Pagination` moves to React, `SearchBar` will have to be updated.
3) Nevertheless, either way `handleSearch` will have to be updated to a Redux action.

In conclusion, yes, we should call `updateTable` here.

ohadlevy commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoComplete.js:
```diff
@@ -44,13 +46,19 @@ class AutoComplete extends React.Component {
       menu,
     } = this.props;
+
+    const tooltip = (<Tooltip id="__clear-tooltip__">{__('Clear')}</Tooltip>);
```
maybe we should give it a unique id in case there are other clear tooltips on the dom?

ohadlevy commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/helpers.js:
```diff
@@ -1,33 +1,20 @@
 import React from 'react';
 import { Menu, MenuItem } from 'react-bootstrap-typeahead';
+import { groupBy } from 'lodash';
```
please use:
```js
import groupBy from 'lodash/groupby'
```
this ensure we only load the parts we need from lodash vs all of it.

amirfefer commented on app/views/common/_searchbar.html.erb:
```diff
@@ -1,21 +1,15 @@
-  documentationUrl: documentation_url("4.1.5Searching")
-}.to_json) %>
+<%= mount_react_component('SearchBar', '.search-bar', {
+    controller: auto_complete_controller_name,
```
Thoughts about moving this hash to a helper? for instance we're about to use `bookmarks` in other components  (i.e #5620)

amirfefer commented on webpack/assets/javascripts/react_app/common/helpers.js:
```diff
@@ -1,11 +1,22 @@
+  });
+};
+
+export const bindMethods = (context, methods) => {
```
patternfly react removed the `bindMethods` implemetioson and moved to `class properties` -  https://github.com/patternfly/patternfly-react/issues/294. @sharvit  any thoughts?
I found this as a cleaner way.

amirfefer commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoComplete.js:
```diff
@@ -0,0 +1,108 @@
+    return (
+      <div>
+        <TypeAheadSelect
+          ref={(ref) => { this._typeahead = ref; }}
```
ref API has been changed in react 16.3
https://reactjs.org/blog/2018/03/29/react-v-16-3.html

amirfefer commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoComplete.js:
```diff
@@ -0,0 +1,108 @@
+      menu,
+    } = this.props;
+
+    const tooltip = (<Tooltip id="__clear-tooltip__">{__('Clear')}</Tooltip>);
```
is there a  reason for a double underscore in the id field?

boaz0 replied:
Yep, I don't see a reason why. 
Why didn't ESLint say anything?

amirfefer commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoComplete.js:
```diff
@@ -0,0 +1,108 @@
+       *  HACK: I had no choice but to call to an inner function,
+       * due to lack of design in react-bootstrap-typeahead.
+        */
+    this._typeahead.getInstance()._showMenu();
```
`showMenu` is not a [public method](https://github.com/ericgio/react-bootstrap-typeahead/blob/master/docs/Methods.md), it might cause issues.

boaz0 replied:
Look at the comments, Ron and I wanted to figure a way to show menu after selecting item. However, `react-typeahead-bootstrap` believes it is a good idea to hide the menu after selecting w/o providing a way to change that behavior.

amirfefer commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoComplete.js:
```diff
@@ -0,0 +1,108 @@
+}
+
+AutoComplete.propTypes = {
+  data: PropTypes.object,
```
could you use PropTypes.shape instead?

amirfefer commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoComplete.js:
```diff
@@ -0,0 +1,108 @@
+
+AutoComplete.defaultProps = {
+  data: {},
+  getResults: () => null,
```
you can use `noop`

amirfefer commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoComplete.js:
```diff
@@ -0,0 +1,108 @@
+AutoComplete.propTypes = {
+  data: PropTypes.object,
+  getResults: PropTypes.func,
+  getInitialState: PropTypes.func,
```
do you use this prop anywhere?

amirfefer commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoCompleteActions.js:
```diff
@@ -0,0 +1,32 @@
+} from './AutoCompleteConstants';
+
+export const getResults = (path, searchQuery, controller, trigger = '') => (dispatch) => {
+  const url = `${path}${URI.encodeQuery(searchQuery)}${trigger === 'Item selected' ? '+' : ''}`;
```
does it make sense to extract it to a `urlBuilder` function?

amirfefer commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoCompleteReducer.js:
```diff
@@ -0,0 +1,59 @@
+export default (state = initialState, action) => {
+  const { payload, type } = action;
+  if (!payload) {
+    return state;
```
the state returns in the default switch case, is it really needed to return it here as well?

amirfefer commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoCompleteSelector.js:
```diff
@@ -0,0 +1,12 @@
+
+export const resultsSelector = createSelector(
+  ({ autocomplete: { results } }) => (results.asMutable ? results.asMutable() : results),
+  results => results,
```
does it need memorization?

amirfefer commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/helpers.js:
```diff
@@ -0,0 +1,32 @@
+  return <Menu id="__auto-complete__">{items}</Menu>;
+};
+
+export const clearSpaces = string => string.trim().replace(/\s\s+/, ' ');
```
This could be in a general helper file

amirfefer commented on webpack/assets/javascripts/react_app/components/SearchBar/index.js:
```diff
@@ -0,0 +1,6 @@
+import { connect } from 'react-redux';
+import SearchBar from './SearchBar';
+
+const mapStateToProps = ({ autocomplete }) => ({ searchQuery: autocomplete.searchQuery });
```
does it need to be connected ?

boaz0 commented:
I am just gonna put [this (downshift)](https://github.com/paypal/downshift) here.
Maybe `react-typeahead-bootstrap` can be replaced by it in pf-react?

boaz0 commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoCompleteActions.js:
```diff
@@ -0,0 +1,50 @@
+    }))
+    .catch(error => dispatch({
+      type: AUTO_COMPLETE_FAILURE,
+      payload: { error, ...items, status: STATUS.ERROR },
```
Wait a second... why aren't you using `ajaxRequestAction`?
Why do you have to pass `STATUS.{ERROR,RESOLVED}` here? Most of the time we're doing this in the reducer.

Ron-Lavi replied:
great question @ripcurld, it's because while using `ajaxRequestAction` in `AUTO_COMPLETE_SUCCESS` `payload` we got the `results` mixed with all of the other items (  e.g: `{0,1,2,3, url, searchQuery, controller...}`), and needed to filter them and return only the results, not the items.

In this action I return the `data` that I've recieved from the `API call` as `results`.
Later in the reducer I just pick them as `payload.results` without the need to filter. :)

sharvit commented on app/views/common/_searchbar.html.erb:
```diff
@@ -1,21 +1,2 @@
-  documentationUrl: documentation_url("4.1.5Searching")
-}.to_json) %>
+<div class="search-bar"></div>
+<%= mount_search_bar(".search-bar"); %>
```
Shouldn't it suppose the based on ID?
```erb
<div class="search-bar" id="search-bar"></div>
<%= mount_search_bar("#search-bar"); %>
```

Ron-Lavi replied:
I think it works with any `querySelector`, 
in my opinion class would be better for cases where there are two `SearchBars` in a single page

sharvit replied:
Can you actually support 2 `SearchBars` in a single page? It won't be weird?
I mostly afraid the `SearchBars` will leak into different places.

sharvit commented on package.json:
```diff
@@ -74,10 +74,11 @@
     "prop-types": "^15.6.0",
     "react": "^16.3.1",
     "react-bootstrap": "^0.32.1",
+    "react-bootstrap-typeahead": "^3.1.5",
```
Do we actually need it? patternfly-react does not deliver it?

Ron-Lavi replied:
`react-bootstrap-typeahead` doesn't expose their inner components,
I am going to push a [PR](https://github.com/patternfly/patternfly-react/pull/497) to Patternfly, and than we won't need to import this library.

sharvit commented on webpack/assets/javascripts/react_app/common/helpers.js:
```diff
@@ -1,11 +1,23 @@
+export const debounceMethods = (context, time, methods) => {
+  methods.forEach((method) => {
+    // eslint-disable-next-line no-param-reassign
+    context[method.name || method] = debounce(context[method.name || method], time || method.time);
```
Could you make it more readable?
```js
const methodName = method.name || method;
const methodTime = time || method.time;
context[methodName] = debounce(context[methodName], methodTime);
```

sharvit replied:
Shouldn't it suppose to be `method.time || time`?

Ron-Lavi replied:
oops! you're right :+1:

sharvit commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/auto-complete.scss:
```diff
@@ -0,0 +1,33 @@
+@import '~patternfly/dist/sass/patternfly/_color-variables.scss';
+@import '~patternfly-react/dist/sass/_type-ahead-select.scss';
```
Does the `_type-ahead-select.scss` actually needed?

Ron-Lavi replied:
Yes, it must have styles.. otherwise things will break.

sharvit replied:
So why doesn't `patternfly-react` import it by default?

Ron-Lavi replied:
removed this line and it works without the import !

sharvit commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/helpers.js:
```diff
@@ -0,0 +1,59 @@
+import SubstringWrapper from '../common/SubstringWrapper';
+import { TRIGGERS } from './AutoCompleteConstants';
+
+export const renderMenu = ({ results, searchQuery }) => {
```
Why is it an `helper` and not a `component`?

Ron-Lavi replied:
Eventually this is a prop for `Autocomplete` 
though I can see why it can become a component.

sharvit commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoCompleteSelector.js:
```diff
@@ -0,0 +1,7 @@
+import { createSelector } from 'reselect';
+import { renderMenu } from './helpers';
+
+export const menuSelector = createSelector(
```
Could you separate it to two selectors?

```js
export const selectState = state => state.autocomplete;

export const selectAutoCompleteMenu = createSelector(
  selectState,
  ({ results, searchQuery }) => renderMenu({ results, searchQuery }),
);
```

sharvit commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/AutoCompleteSelector.js:
```diff
@@ -0,0 +1,7 @@
+
+export const menuSelector = createSelector(
+  ({ autocomplete: { results, searchQuery } }) => ({ results, searchQuery }),
+  menuProps => renderMenu(menuProps),
```
Does it make sense to have a selector that return component?
Shouldn't it return a json that a component receive?

sharvit commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/helpers.js:
```diff
@@ -0,0 +1,59 @@
+
+export const clearSpaces = string => string.trim().replace(/\s\s+/, ' ');
+
+export const getActionsHistory = (state) => {
```
Shouldn't it suppose to be a Selector?

sharvit commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/helpers.js:
```diff
@@ -0,0 +1,59 @@
+  return actionsHistory;
+};
+
+export const getLastAction = (state) => {
```
Shouldn't it suppose to be a Selector?

sharvit commented on webpack/assets/javascripts/react_app/components/AutoCompleteSearch/index.js:
```diff
@@ -0,0 +1,118 @@
+  constructor(props) {
+    super(props);
+    bindMethods(this, ['handleClear', 'handleInputChange', 'handleResultsChange', 'handleInputFocus', 'getResults']);
+    debounceMethods(this, 250, ['handleInputChange', 'handleLoading', { name: 'handleInputFocus', time: 110 }]);
```
Wondering... Would it be easier to follow as well?

```js
debounceMethods(this, 250, ['handleInputChange', 'handleLoading']);
debounceMethods(this, 110, ['handleInputFocus']);
```

Ron-Lavi replied:
I wanted to give an option in `debounceMethods` to control 
each function debounce-time without calling the function couple of times.
e.g:
`debounceMethods(this, 250, [`
    `'handleInputChange',` 
   `'handleLoading',` 
    `{ name: 'handleInputFocus', time: 110 },`
    `{ name: 'handleClear', time: 500 },`
`]);`

sharvit replied:
I can see it useful when each method has different timing, but IMO in your case when you got only one exception, it easier to read it with two calls.

sharvit commented on webpack/assets/javascripts/react_app/components/SearchBar/SearchBar.js:
```diff
@@ -0,0 +1,44 @@
+};
+
+const SearchBar = ({
+  data, autocomplete, resetData, getResults,
```
It doesn't match the propTypes

sharvit commented on webpack/assets/javascripts/react_app/components/SearchBar/SearchBar.js:
```diff
@@ -0,0 +1,44 @@
+  <div className="input-group">
+      <AutoComplete
+        data={ { ...data.autocomplete, controller: data.controller } }
+        {...{ ...autocomplete, resetData, getResults }}
```
Could it be?
```js
resetData={resetData}
getResults={getResults}
{ ...autocomplete}
```

sharvit commented on webpack/assets/javascripts/react_app/components/SearchBar/index.js:
```diff
@@ -0,0 +1,23 @@
+
+const mapStateToProps = state => ({
+  autocomplete: {
+    ...state.autocomplete,
```
Could you destruct individual props? It might be dungarees to pass all the state with no control over it.

sharvit commented on webpack/assets/javascripts/react_app/components/SearchButton/index.js:
```diff
@@ -0,0 +1,11 @@
+import React from 'react';
+import { Button, Col, Icon } from 'patternfly-react';
+
+const SearchButton = props => (
```
Shouldn't it suppose to be AutoCompleteSearchButton?
* mostly because `className="autocomplete-search-btn-text"`

sharvit reviewed (changes requested):
Nice progress @Laviro :+1: 

Left some comments

amirfefer commented:
There are some test failures:
from travis:
```
SearchBar integration test › should flow
  Enzyme Internal Error: unknown node with tag 12
```
Plus some related foreman integration tests fails.

Ron-Lavi commented:
thanks @amirfefer, 
React in version 16.3 introduces new Context API. 
That API comes with two new node types (12 and 13).
These node types are not supported and therefore enzyme-adapter-react-16 is throwing an error.

Until enzyme will release their solution: https://github.com/airbnb/enzyme/issues/1553 
to support some of react 16 features, I am unabling the test.

How ever, all of the tests pass while using React 16.2,
where the coverage is 100%.
![selection_034](https://user-images.githubusercontent.com/26363699/43722877-ffdc9854-999e-11e8-9d84-13dc0665bfe2.png)
![selection_035](https://user-images.githubusercontent.com/26363699/43722886-036d4734-999f-11e8-94ef-dba5e3df5db0.png)

ohadlevy commented on webpack/assets/javascripts/react_app/common/helpers.js:
```diff
@@ -1,11 +1,28 @@
+
 export const noop = Function.prototype; // empty function
 
+export const clearSpaces = string => string.trim().replace(/\s\s+/, ' ');
```
do we need both trip and replace?

Ron-Lavi replied:
The trim method only removes whitespace from both ends of a string, but doesn't replace double spacing inside it.

ohadlevy commented on webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteReducer.js:
```diff
@@ -0,0 +1,72 @@
+        results,
+        actionsHistory,
+      });
+    case AUTO_COMPLETE_INITIAL_QUERY:
```
this still feels like an extra action? does it make sense to combine this event with other events? (such as used_query or success)?

Ron-Lavi replied:
We can switch it by the `getResults` action, 
but this means that an API call will be made when the page loads which may cause the loading spinner to be seen, something we wanted to avoid earlier.

the `updateSearchQuery` action, updates the store with the `query` and `controller`, that comes initially from the server side as the component is mounted, without loading the results.

Ron-Lavi replied:
Removed the AUTO_COMPLETE_INITIAL_QUERY action,
now using AUTO_COMPLETE_SUCCESS

boaz0 commented on webpack/assets/javascripts/react_app/components/AutoComplete/__tests__/__snapshots__/AutoCompleteReducer.test.js.snap:
```diff
@@ -16,7 +14,7 @@ Object {
 exports[`AutoComplete reducer should handle AUTO_COMPLETE_INITIAL_QUERY 1`] = `
 Object {
-  "controller": "models",
+  "controller": null,
```
is that supposed to be `null`?

Ron-Lavi replied:
Some tests have failed after the refactor, this one should be deleted :)

boaz0 commented on webpack/assets/javascripts/react_app/components/AutoComplete/__tests__/__snapshots__/AutoCompleteReducer.test.js.snap:
```diff
@@ -52,25 +50,18 @@ Object {
 exports[`AutoComplete reducer should handle AUTO_COMPLETE_SUCCESS 1`] = `
 Object {
-  "controller": null,
+  "controller": undefined,
```
I think you have a bug - on success I would imagine `controller` be a string and neither `null` nor `undefined`.

boaz0 commented on webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteReducer.js:
```diff
@@ -0,0 +1,57 @@
+});
+
+export default (state = initialState, action) => {
+  const { payload, type } = action;
```
Maybe:
```js
const { payload = {}, type } = action
```
and remove the `|| {};` from
```js
const {
  ...
} = payload || {};
```

Ron-Lavi replied:
Thanks, looks better !

boaz0 commented on webpack/assets/javascripts/react_app/components/AutoComplete/__tests__/AutoCompleteReducer.test.js:
```diff
@@ -0,0 +1,54 @@
+  AUTO_COMPLETE_FAILURE,
+  AUTO_COMPLETE_RESET,
+  AUTO_COMPLETE_USED_QUERY,
+  AUTO_COMPLETE_INITIAL_QUERY,
```
Remove `AUTO_COMPLETE_INITIAL_QUERY`.

boaz0 replied:
Also, you should add tests to `TRIGGERS`, shouldn't ya?

boaz0 commented on webpack/assets/javascripts/react_app/components/AutoComplete/__tests__/AutoCompleteReducer.test.js:
```diff
@@ -0,0 +1,54 @@
+
+const fixtures = {
+  'should return the initial state': {},
+  'should handle AUTO_COMPLETE_INITIAL_QUERY': {
```
Is this test relevant?

boaz0 commented on webpack/assets/javascripts/react_app/components/SearchBar/index.js:
```diff
@@ -0,0 +1,23 @@
+import reducer from '../AutoComplete/AutoCompleteReducer';
+import SearchBar from './SearchBar';
+
+
```
can you remove this extra new line

boaz0 commented on webpack/assets/javascripts/react_app/components/componentRegistry.js:
```diff
@@ -58,16 +58,16 @@ const componentRegistry = {
-  { name: 'StorageContainer', type: StorageContainer },
+  { name: 'PowerStatus', type: PowerStatus },
   { name: 'PasswordStrength', type: PasswordStrength },
-  { name: 'BreadcrumbBar', type: BreadcrumbBar },
```
I am puzzled. Why did you change the order of `coreComponents` alphabetically?

Ron-Lavi commented:
-- Update --
Enzyme & enzyme-adapter-react-16.3 have just released a new versions which fixed the bugs.
Now the tests pass and coverage is good again.

glekner commented on webpack/assets/javascripts/react_app/components/SearchBar/SearchBar.js:
```diff
@@ -0,0 +1,84 @@
+  </div>
+);
+
+SearchBar.propTyoes = {
```
typo: 'propTypes'

glekner commented on webpack/assets/javascripts/react_app/components/AutoComplete/components/AutoCompleteClearButton.js:
```diff
@@ -0,0 +1,20 @@
+
+
+const AutoCompleteClearButton = ({ onClear }) => {
+  const tooltip = (<Tooltip id={UUID()}>{__('Clear')}</Tooltip>);
```
Do we render this button more than once?
why not just `<Tooltip id="clearbutton" />`

Ron-Lavi replied:
There might be other tooltips in the DOM, so we want to use a unique id.
This is how it used to be before: https://github.com/theforeman/foreman/pull/5771/commits/0817d0318143c6a2fa3db529e1127e7d8ecf1f16#diff-58cdf6e5ce7e5afcd2cd978a1bb22067

lizagilman replied:
maybe have the id from a prop? with some default value

glekner reviewed (commented):
Good job @Laviro !

ohadlevy commented:
a few things I've noticed:

1. writing a search term and hitting enter does not send the query
2. when in paginated view (e.g. page=2) and changing the search term, the page is not reset (it should be removed or reset back to page=1)
3. when starting to write a search term that is not valid (e.g. "x ~") the UI does not show an error, it only show it after clicking on search.
4. if the search term is long, its very hard to navigate to the suggestions 
![image](https://user-images.githubusercontent.com/20167/44148674-f0f4f48a-a0a1-11e8-80a9-ad26fb6c33da.png)

Ron-Lavi commented:
Thanks @ohadlevy, these are some really good points. going to fix them.

boaz0 commented:
@Laviro @tbrisker 
I see this error when running foreman tests:

```shell
ModuleNotFoundError: Module not found: Error:
Can't resolve '@babel/runtime/helpers/es6/inheritsLoose' in '/var/lib/workspace/workspace
/test_develop_pr_core/database/postgresql/ruby/2.4/slave
/fast/node_modules/react-bootstrap/es'
```

Ron-Lavi commented:
@ohadlevy I have fixed the issues that you've mentioned above,
and added a shortcut to focus on the search input by hitting `forward-slash (/)`,just like in Github.

Going to work on some more tests for the new additions.

Ron-Lavi commented:
Feel free to check out the new [storybook](https://laviro.github.io/foreman/?selectedKind=Components%2FSearchBar&selectedStory=Search%20Bar&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Factions%2Factions-panel).

ares commented on app/helpers/search_bar_helper.rb:
```diff
@@ -0,0 +1,16 @@
+module SearchBarHelper
+  def mount_search_bar(id)
```
it would be great to have an option here to disable bookmarks and the `/` shortcut accesssor

e.g. on filters page or discovery rules form we don't want to have bookmarks as part of the search component

also if we have 2 or more search boxes on one page, the shortcut would be confusing and it should probably be always used only to activate the top level search

Ron-Lavi replied:
Thanks, that's a really great idea !

lizagilman commented:
[test foreman]

Ron-Lavi commented:
![screenshot_20180823-191521_whatsapp](https://user-images.githubusercontent.com/26363699/44538106-1abdf100-a709-11e8-8074-c44cef38066d.jpg)

Seems that all test have passed, except for something unrelated,
does anyone has more comments? 
Thanks

ares commented on app/helpers/search_bar_helper.rb:
```diff
@@ -0,0 +1,21 @@
+module SearchBarHelper
+  def mount_search_bar(id, use_shortcuts = true, use_bookmarks = true)
```
I'd prefer to either use hash as options or better, keyword arguments, basically that only means to change changing ` use_shortcuts = true, use_bookmarks = true` to ` use_shortcuts: true, use_bookmarks: true`. And then when you call the method, the key name must be specified. The difference/benefit is that when we add more options, it does not change the method prototype too much, the arguments position does not matter.

And since this is a helper that will be used from more places, I'd prefer to keep it easily extendable

Another thing that I just realized, this assumes the controller can determine the url based on its name. While it might not be blocking this PR, it would be great to be able to pass the url as another param (might be a nice little enhancement for other who wants to get hands dirty). What's more important I think though is the params[:search] which is hardcoded now to searchQuery. On other pages, I might want to link it to different parameter.

ares commented on test/integration/config_report_js_test.rb:
```diff
@@ -1,9 +1,26 @@
+    visit config_reports_path
+    click_link(@report.host.fqdn)
+    sleep 3
+    # TODO check why it doesn't find the search id.
```
when will this TODO be resolved? :-) I think we shouldn't merge code with TODOs

ares commented:
there are several TODOs in tests, should those be addressed first?

Ron-Lavi commented:
Thanks for your review @ares !

lizagilman commented on .gitignore:
```diff
@@ -48,3 +49,4 @@ public/webpack
 package-lock.json
 npm-debug.log
 .vscode
+postgresql/*
```
did you add gitignore on purpose..?

boaz0 commented on test/integration/config_report_js_test.rb:
```diff
@@ -1,9 +1,25 @@
+  test "reports for host" do
+    visit config_reports_path
+    click_link(@report.host.fqdn)
+    sleep 3
```
I don't think you need to use `sleep` here. Instead of `sleep` maybe put `search = find('.rbt-input-main')` and assert on `search.value`.

boaz0 commented on test/integration/fact_value_js_test.rb:
```diff
@@ -0,0 +1,32 @@
+    within(:xpath, "//tr[contains(.,'#{@fact_name.name}')]") do
+      click_link(@host.fqdn)
+    end
+    sleep 3
```
if you remove `sleep 3` and run the test on your machine, does it fail there or it only fails on Jenkins?
Because I am puzzled.

Ron-Lavi replied:
if I remove `sleep` it fails on my machine

boaz0 commented on webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js:
```diff
@@ -0,0 +1,97 @@
+  const loadNextResults = trigger === TRIGGERS.ITEM_SELECT ? '+' : '';
+  const url = `${path}${URI.encodeQuery(searchQuery)}${loadNextResults}`;
+  // to avoid an extra API call, find if this search query was already used.
+  const usedQueryResults = queryCache[controller] && queryCache[controller][searchQuery];
```
There is a problem with the caching, I think.

Scenario:
1) open two tabs:
  a) hosts index page
  b) hosts create page
2) search a specific host name in the first tab
3) create a new host in the second one
4) in the first tab searching for the host that I just created by name is not shown...

Ron-Lavi replied:
yeah, you absolutely right.. I will think of something..

boaz0 commented on webpack/assets/javascripts/react_app/components/AutoComplete/index.js:
```diff
@@ -0,0 +1,200 @@
+    }
+  }
+
+  undableHTMLAutocomplete() {
```
typo? `undableHTMLAutocomplete` -> `unableHTMLAutocomplete`?

Ron-Lavi replied:
thanks

boaz0 reviewed (commented):
:handshake:

boaz0 commented:
@Laviro there are still 14 failed tests in `foreman`. Do you mind checking why?
Let me know if you need help.

Ron-Lavi commented:
@ripcurld Thanks for the review ! 
yeah this test sometimes fail, and sometime doesn't.. I guess it is because I used `sleep`..

boaz0 commented:
[test foreman]

sharvit commented:
Some comments:

1. I don't really like the new design when the background is gray when not focused. It feels to me like the field is disabled and I shouldn't use it.

2. When I click the search button and getting the search results, the search field value is dispersing. Is that the desirable experience? It feel weird to me anyway.
    And when I click `bookmark this search` the search doesn't automatically copied to the bookmarks form.

boaz0 commented:
[test foreman]

Ron-Lavi commented:
@sharvit Thanks for the review, I will check why it doesn't work - it supposed to.
about the design - I tried to add a behavior like GitHub or Gmail has, as the search looks off until you focus on it by `slash` or `click`

Ron-Lavi commented:
@sharvit @ripcurld Yesterday I have changed the props order and it seems 
that some data from the server while mounting, overrides the components changing props, 
I fixed that and now it works :)

Ron-Lavi commented:
@ripcurld I removed the `sleep` method from `capybara` tests and instead use `has_selector?`

Ron-Lavi commented:
@boaz1337 Decided to drop the query-caching for this PR, 
and maybe add it as a follow up. :)

boaz0 commented on webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js:
```diff
@@ -0,0 +1,76 @@
+  const loadNextResults = trigger === TRIGGERS.ITEM_SELECT ? '+' : '';
+  const url = `${path}${URI.encodeQuery(searchQuery)}${loadNextResults}`;
+
+  // Starting a request
```
unnecessary comment.

boaz0 commented on webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js:
```diff
@@ -0,0 +1,76 @@
+    },
+  });
+
+  // Making an API call
```
unnecessary comment 2

boaz0 commented on webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js:
```diff
@@ -0,0 +1,76 @@
+  return API.get(url)
+    .then(({ data }) => {
+      const { error } = data[0] || {};
+      // The API can return errors too.
```
unnecessary comment 3

boaz0 commented on webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js:
```diff
@@ -0,0 +1,76 @@
+      const { error } = data[0] || {};
+      // The API can return errors too.
+      if (error) {
+        throw error;
```
I don't know what is better but I like doing `Promise.reject(error)`.
@amirfefer @sharvit any preferences? Otherwise, lets keep `throw error`.

boaz0 commented on webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js:
```diff
@@ -0,0 +1,76 @@
+        payload: {
+          controller,
+          results: data,
+          searchQuery,
```
I believe data has `searchQuery` in its body.

boaz0 replied:
in other words, I believe we can use [`ajaxRequestAction`](https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/redux/actions/common/index.js#L3) here.

Ron-Lavi replied:
The data returned from the request is the results in array, e.g:
![selection_050](https://user-images.githubusercontent.com/26363699/44723810-b5d41380-aad9-11e8-99e7-7968c1b66b28.png)
That's why I needed to make some adjustments that the `ajaxRequestAction` doesn't support, 
I also needed to handle API errors that are resolved after the API request.

> I believe data has searchQuery in its body.

not sure that I understand what did you mean.

boaz0 replied:
My bad - I thought it returns `search`, `total`, and etc like in `api/v2/`

Laviro commented:
[test upgrade]

Laviro commented:
[test foreman]

Laviro commented:
[test foreman]

Laviro commented:
[test foreman]

amirfefer reviewed (commented):
Looks like tests failures are unrelated.

boaz0 commented:
I agree with @amirfefer - it seems like an unrelated failure.
Maybe doing `pull --rebase` will fix it, I don't know.

// cc @tbrisker

Ron-Lavi commented:
@amirfefer @boaz1337 seem to appear in other recent PRs

Ron-Lavi commented:
@rohoover Do you have any thoughts about the design of the [SearchBar](https://laviro.github.io/foreman/?selectedKind=Components%2FSearchBar&selectedStory=Search%20Bar&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Factions%2Factions-panel)?

@ares Any improvements to the ruby part of this PR?

@amirfefer @sharvit Can you review?

thanks :)

Rohoover commented:
@Laviro 

A few thoughts:

1) I'm glad you're looking at this. Getting consistency with our search bar is important. I've talked to a number of customers, some of whom have noticed how our search is inconsistent both in implementation and styling.

2) Can we get the default filter input area to be white, not grey? It looks like it's disabled with grey and the contrast between the grey lettering and grey background is probably not sufficient enough for those with visual impairment. This would make it compliant with the current PF search styling.

3) The forward slash icon located in search bar. I'm not keen on it's placement there. It's not clickable whereas the "x" is. It's also an unusual style deviation from core PF. I understand the intent, but perhaps this is best described in the documentation. 

4) Can we allow enter to initiate a search when the focus is on the filter input box (not type ahead dropdown) without having to tab to search?

As a note, the PF design team is looking at Filtering and Search now. This document is just a proposal but I like the idea of it being componentized allowing product team flexibility.

Ron-Lavi commented:
@Rohoover thanks !!

> our search is inconsistent both in implementation and styling.

Here we have used it more as an autocomplete I guess..
We can also add it to Patternfly-React as Autocomplete or as another example of TypeAheadSelect.

> Can we get the default filter input area to be white, not grey?

yes, I understand the need.. 

> The forward slash icon located in search bar. 

I will add documentation in storybook
> allow enter to initiate a search when the focus is on the filter input box

Currently `Enter` will cause a search in case there is no dropdown-item selected.

lizagilman commented:
@Laviro  

@Rohoover, are you suggesting we should not keep the forward slash here?

But if we can keep it, how about -

1. Moving it to the left
and / or
2. Adding a tooltip with explanation. something like "press / to focus on search". 

Btw @amirfefer I think it could be a good candidate for the [ui tour component](https://github.com/theforeman/foreman/pull/5946)

![image](https://user-images.githubusercontent.com/15361156/44953959-2dbc8800-aea4-11e8-9f3d-f4e2d7737dc5.png)

![image](https://user-images.githubusercontent.com/15361156/44954461-411f2180-aeab-11e8-9065-76bed0491f2b.png)

![image](https://user-images.githubusercontent.com/15361156/44953986-5fcdea00-aea4-11e8-9de9-9257a0e6ffd5.png)

![image](https://user-images.githubusercontent.com/15361156/44954012-c4894480-aea4-11e8-8899-af7677ac4a9a.png)

@Laviro , when typing a filter field and then an operator, there are suggestions (as expected), but if you continue typing (the value), it is not auto-completed. Is this a bug or a feature? :)

Ron-Lavi commented:
Great eye @lizagilman !! I fixed that feature / bug.
@Rohoover added tooltip for slash shortcut 
and changed input background back to white.

lizagilman commented:
Thanks @Laviro 
Can you add a screenshot of the new changes?
And tests should be ok now

Ron-Lavi commented:
@lizagilman Here you can see the changes, and also you can always try it yourself in the [storybook](https://laviro.github.io/foreman/?selectedKind=Components%2FSearchBar&selectedStory=Search%20Bar&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Factions%2Factions-panel)
![sep 2 2018 8_25 pm](https://user-images.githubusercontent.com/26363699/44959684-1957aa00-aefb-11e8-89bd-91344ab656d3.gif)

Made rebase, and now all test have passed.

sharvit commented on app/helpers/search_bar_helper.rb:
```diff
@@ -0,0 +1,27 @@
+    options = {
+      use_key_shortcuts: true,
+      use_bookmarks: true,
+      url: "#{auto_complete_controller_name}/auto_complete_search?search=",
```
What is the reason we need the `?search=` as a suffix? Shouldn't we add this param with javascript?

Ron-Lavi replied:
this is the `path` for the `API request` so people can change it if they use it elsewhere,
do you think I should rename it to be clearer?

sharvit replied:
So why can't it be?
```ruby
url: "#{auto_complete_controller_name}/auto_complete_search",
```

Ron-Lavi replied:
in case some API doesn't work with `?search`, I let the consumer to choose the API path

ohadlevy replied:
I agree with Avi its not required, but I would not block on this :)

Ron-Lavi replied:
As it seem in the [encodeQuery from urijs](https://medialize.github.io/URI.js/docs.html#static-encodeQuery), it doesn't add a key and only encode the given value.

I modify the API path only by adding a value to it..

tbrisker replied:
you could use [`URI.buildQuery`](https://medialize.github.io/URI.js/docs.html#static-buildQuery) instead. 
Also, are there any specific examples of this path not working? I would prefer only adding an option when it is needed, this function already takes a whole bunch of them

sharvit commented on webpack/assets/javascripts/react_app/common/helpers.js:
```diff
@@ -1,17 +1,41 @@
+import debounce from 'lodash/debounce';
+
+export const debounceMethods = (context, time, methods) => {
```
Nice Work with those helpers functions :+1: 

Can you please add JSdoc style comments?
One of the UI meeting goals was to add those comments to infrastructure methods and we will try to generate docs from it.

Example:
https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/common/testHelpers.js#L38-L43

sharvit commented on webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js:
```diff
@@ -0,0 +1,75 @@
+} from './AutoCompleteConstants';
+import { objectDeepTrim } from '../../common/helpers';
+
+export const getResults =
```
If you want to practice `async/await` syntax, I think it will make this method looks better :)

sharvit commented on webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js:
```diff
@@ -0,0 +1,75 @@
+    }));
+};
+
+export const resetData = controller => (dispatch) => {
```
Can remove the Curly braces.

sharvit commented on webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js:
```diff
@@ -0,0 +1,75 @@
+  });
+};
+
+export const initialUpdate = (searchQuery, controller) => (dispatch) => {
```
Can remove the Curly braces.

sharvit commented on webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteReducer.js:
```diff
@@ -0,0 +1,58 @@
+      status,
+      trigger,
+    } = {}, type,
+  } = action;
```
Just small cosmetic but I think it will look better as:
```js
  const {
     type,
     payload: {
       controller,
       error,
       results,
       searchQuery,
       status,
       trigger,
     } = {},
   } = action;
```

sharvit commented on webpack/assets/javascripts/react_app/components/AutoComplete/__tests__/AutoComplete.test.js:
```diff
@@ -0,0 +1,124 @@
+import React from 'react';
```
Great test :+1:

Ron-Lavi replied:
Thanks :)))

sharvit commented on webpack/assets/javascripts/react_app/components/AutoComplete/components/AutoCompleteFocusShortcut.js:
```diff
@@ -0,0 +1,25 @@
+import { OverlayTrigger, Tooltip } from 'patternfly-react';
+
+const AutoCompleteFocusShortcut = ({ useKeyShortcuts, tooltipID }) => {
+  const tooltip = useKeyShortcuts ? (<Tooltip id={tooltipID}>{__("Press ' / ' to focus on search")}</Tooltip>) : null;
```
You could use:

```js
const tooltip = useKeyShortcuts && (<Tooltip id={tooltipID}>{__("Press ' / ' to focus on search")}</Tooltip>);
```

sharvit commented on webpack/assets/javascripts/react_app/components/AutoComplete/components/AutoCompleteFocusShortcut.js:
```diff
@@ -0,0 +1,25 @@
+  const tooltip = useKeyShortcuts ? (<Tooltip id={tooltipID}>{__("Press ' / ' to focus on search")}</Tooltip>) : null;
+  return (
+  <OverlayTrigger overlay={tooltip} placement="top" trigger={['hover', 'focus']}>
+    <span className={`autocomplete-focus-shortcut ${!useKeyShortcuts ? 'hide' : ''}`}>/</span>
```
Could you use [classnames](https://github.com/JedWatson/classnames)?

sharvit commented on webpack/assets/javascripts/react_app/components/AutoComplete/components/AutoCompleteMenu.js:
```diff
@@ -0,0 +1,42 @@
+const AutoCompleteMenu = ({ results, menuProps }) => {
+  let itemIndex = 0;
+  const grouped = groupBy(results, r => r.category);
+  const items = Object.keys(grouped).sort().map(category => [
```
Would it make more sense to return a `React.Fragment` instead `Array`?

Ron-Lavi replied:
Yeah, I agree :)

sharvit commented on webpack/assets/javascripts/react_app/components/AutoComplete/components/AutoCompleteMenu.js:
```diff
@@ -0,0 +1,42 @@
+const { Menu, MenuItem } = TypeAheadSelect;
+
+const AutoCompleteMenu = ({ results, menuProps }) => {
+  let itemIndex = 0;
```
I don't think you actually need to manage you own index.
The `map` function can give you the current index together with the value.

```js
Object.keys(grouped).sort().map((category, index) => ... );
```

Ron-Lavi replied:
The only problem is that on every new category,
the index will be reset to 0.
this means that there would be to children with the same key, 
and there would be an error. :(

sharvit replied:
What about?
```jsx
<MenuItem key=`${category}-${resultIndex}` />
```

Ron-Lavi replied:
Still doesn't work as we need to count the menuItems position,
e.g. : ` <MenuItem key={itemIndex} option={result.label} position={itemIndex}>`

sharvit commented on webpack/assets/javascripts/react_app/components/SearchBar/__tests__/integration.test.js:
```diff
@@ -0,0 +1,98 @@
+import React from 'react';
```
Great integration test :+1:

Ron-Lavi replied:
Thanks!!

sharvit reviewed (changes requested):
Really awesome work @Laviro :+1: 

I believe your code is going to be a great example for other developers about how to do things right.
High code quality, good attention to details and great tests :+1:

Rohoover commented:
I like the repositioning of the slash as well as the tooltip (which I think
is required).

On Mon, Sep 3, 2018 at 7:55 AM Ron Lavi <notifications@github.com> wrote:

> *@Laviro* commented on this pull request.
> ------------------------------
>
> In
> webpack/assets/javascripts/react_app/components/AutoComplete/components/AutoCompleteMenu.js
> <https://github.com/theforeman/foreman/pull/5771#discussion_r214665072>:
>
> > @@ -0,0 +1,42 @@
> +import React from 'react';
> +import PropTypes from 'prop-types';
> +import groupBy from 'lodash/groupBy';
> +import { TypeAheadSelect } from 'patternfly-react';
> +import SubstringWrapper from '../../common/SubstringWrapper';
> +
> +const { Menu, MenuItem } = TypeAheadSelect;
> +
> +const AutoCompleteMenu = ({ results, menuProps }) => {
> +  let itemIndex = 0;
> +  const grouped = groupBy(results, r => r.category);
> +  const items = Object.keys(grouped).sort().map(category => [
>
> Yeah, I agree :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/theforeman/foreman/pull/5771#discussion_r214665072>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ARa5VY5npkMTLAwfVmbbXbZJgzPVU3noks5uXRiogaJpZM4VCnnr>
> .
>
-- 

ROXANNE HOOVER

VISUAL & INTERACTION DESIGN

Red Hat

<https://www.redhat.com/>

rohoover@redhat.com    M: 919-609-5300
<https://red.ht/sig>
TRIED. TESTED. TRUSTED. <https://redhat.com/trusted>

Ron-Lavi commented:
Thanks @sharvit for the great review!!!
@Rohoover not sure I understand, do you like the current design?

sharvit reviewed (dismissed):
LGTM! Thanks @Laviro :+1:

ohadlevy commented on webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js:
```diff
@@ -0,0 +1,75 @@
+  });
+
+
+export const initialUpdate = (searchQuery, controller) => dispatch =>
```
nitpick: extra whitespace above this line, in general I would recommend to use prettier / eslint to avoid many of these issues.

ohadlevy commented on webpack/assets/javascripts/react_app/components/AutoComplete/components/AutoCompleteSearchButton.js:
```diff
@@ -0,0 +1,22 @@
+import { Button, Col, Icon } from 'patternfly-react';
+
+const SearchButton = ({ className, ...props }) => (
+    <Button
```
nitpick: 4 spaces instead of 2

ohadlevy commented on webpack/assets/javascripts/react_app/components/AutoComplete/components/AutoCompleteFocusShortcut.js:
```diff
@@ -0,0 +1,27 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import UUID from 'uuid/v1';
+import cx from 'classnames';
```
classnames is not in our package.json, it should included.

Laviro replied:
@sharvit had suggested we should use it :)

Ron-Lavi replied:
Added :)

ohadlevy reviewed (changes requested):
i had a run of prettier/eslint, resulting with https://gist.github.com/ohadlevy/0e0a380b6ad4780cd5ec314d360bb8bd

ohadlevy commented:
another small issue i saw, is that after operators are selected, spaces are not added to the query, e.g. 
![recorded](https://user-images.githubusercontent.com/20167/45015456-730ac200-b02a-11e8-8192-8464ca6cc9f7.gif)

Ron-Lavi commented:
@ohadlevy Thanks for the review!!
I used `prettier` now in my code and it fixed the small problems,
and I solved the problem you described:
![sep 4 2018 11_41 am](https://user-images.githubusercontent.com/26363699/45020405-8f612b80-b037-11e8-8642-0d9fcbed5d3b.gif)

ohadlevy commented on package.json:
```diff
@@ -78,7 +79,7 @@
     "multiselect": "~0.9.12",
     "patternfly": "^3.42.0",
-    "patternfly-react": "2.5.1",
+    "patternfly-react": "^2.11.3",
```
@ekohl can you confirm these changes? thanks

Ron-Lavi replied:
any decision ? :)

ohadlevy replied:
added packaging pr at https://github.com/theforeman/foreman-packaging/pull/2979

Rohoover commented:
@Laviro 

I like this one:

![screen shot 2018-09-04 at 9 03 14 am](https://user-images.githubusercontent.com/18266453/45033067-6e8ddb80-b021-11e8-9465-b55857a452db.png)

ohadlevy commented:
I found a fairly minor bug ( that I'm ok with fixing later) - it seems that the filtering feature of type ahead is active before the auto completer response from the server, leading to a "jump" in the results, see:
![recorded](https://user-images.githubusercontent.com/20167/45034889-576ede00-b061-11e8-8fb3-bedec2caf0e1.gif)

Ron-Lavi commented:
@Rohoover Updated the design !
and added support for scenarios when the consumer doesn't want to use key-shortcuts.

ohadlevy commented:
[test foreman]

ohadlevy commented:
can you just rebase please? thanks

On Wed, Sep 5, 2018 at 11:00 AM Ron Lavi <notifications@github.com> wrote:

> *@Laviro* commented on this pull request.
> ------------------------------
>
> In
> webpack/assets/javascripts/react_app/components/AutoComplete/components/AutoCompleteFocusShortcut.js
> <https://github.com/theforeman/foreman/pull/5771#discussion_r215169749>:
>
> > @@ -0,0 +1,27 @@
> +import React from 'react';
> +import PropTypes from 'prop-types';
> +import UUID from 'uuid/v1';
> +import cx from 'classnames';
>
> Added :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/theforeman/foreman/pull/5771#discussion_r215169749>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABOx7t9WLsAnbGUnEmE1wc-DCSFHF4bks5uX4SEgaJpZM4VCnnr>
> .
>

tbrisker commented on app/helpers/search_bar_helper.rb:
```diff
@@ -0,0 +1,27 @@
+module SearchBarHelper
+  def mount_search_bar(
+    id,
+    options = {
```
In ruby, unlike JS, if you pass an option hash it will override this default one so you would have to pass all of the options. instead, you can use [kwargs](https://robots.thoughtbot.com/ruby-2-keyword-arguments) to allow setting defaults for any option not provided by the user.

Ron-Lavi commented:
@ohadlevy rebased

tbrisker commented on test/integration/audit_js_test.rb:
```diff
@@ -1,6 +1,6 @@
 require 'integration_test_helper'
 
-class AuditJSTest < IntegrationTestWithJavascript
+class AuditIntegrationJSTest < IntegrationTestWithJavascript
```
This should remain AuditJSTest

tbrisker commented on test/integration/architecture_js_test.rb:
```diff
@@ -0,0 +1,7 @@
+require 'integration_test_helper'
+
+class ArchitectureIntegrationJSTest < IntegrationTestWithJavascript
```
This should be ArchitectureJSTest

tbrisker commented on webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js:
```diff
@@ -0,0 +1,82 @@
+
+export const getResults = (path, searchQuery, controller, trigger) => (dispatch) => {
+  const loadNextResults = trigger === TRIGGERS.ITEM_SELECT ? '+' : '';
+  const url = `${path}${URI.encodeQuery(searchQuery)}${loadNextResults}`;
```
why not build the whole url with uri?

tbrisker reviewed (changes requested):
One main comment regarding the usage of parameters in ruby. 
Also commented on a couple, but js test classes should be named `WhateverJSTest`, not `WhateverIntegrationJSTest`.
Didn't review all of the react code as I trust others have already gone over it more thoroughly.

Ron-Lavi commented:
sorry for deleting your comments by mistake while pushing @marec & @ohadlevy.

- odd snapshot: I forgot to update the actions test, now it looks good.
- passing `url` as an option in `search_bar_helper`: 
Just wanted to be sure.. 
can the API path be different in the filter form? do we use another param instead of `search`?

thanks

ares commented:
> Just wanted to be sure..
can the API path be different in the filter form? do we use another param instead of search?

yes - see https://github.com/theforeman/foreman/blob/develop/app/views/filters/_form.html.erb#L57-L60 and https://github.com/theforeman/foreman/blob/develop/app/helpers/filters_helper.rb#L2-L17

plugins can also override these paths via FiltersHelperOverrides, I think one example would be foreman_openscap

it's always `search` param though

Ron-Lavi commented:
@tbrisker, @ares - moved to use kwargs in the ruby helper and added a `url` param.

ohadlevy commented on package.json:
```diff
@@ -66,6 +66,7 @@
     "babel-polyfill": "^6.26.0",
     "bootstrap-sass": "^3.3.7",
     "brace": "^0.10.0",
+    "classnames": "^2.2.6",
```
we already use 2.2.5 in our current packaging it seems, if you downgrade it we can avoid updating that.

ohadlevy commented:
@tbrisker can you have another look please?

tbrisker commented on app/helpers/search_bar_helper.rb:
```diff
@@ -0,0 +1,25 @@
+      documentationUrl: documentation_url("4.1.5Searching")
+    }
+    mount_react_component('SearchBar', "##{id}", {
+      controller: controller || auto_complete_controller_name,
```
You can set the default value in the kwargs, for example:
```ruby
def mount_search_bar(id:, controller: auto_complete_controller_name, url: "#{auto_complete_controller_name}/auto_complete_search", search_query: params[:search], use_bookmarks: true, use_key_shortcuts: true)

Ron-Lavi replied:
Thanks @tbrisker, I tried this before but it didn't work, though it might be the way I consumed this function, looks much better now !

tbrisker commented on app/helpers/search_bar_helper.rb:
```diff
@@ -0,0 +1,25 @@
+      url: api_bookmarks_path,
+      canCreate: authorizer.can?(:create_bookmarks),
+      documentationUrl: documentation_url("4.1.5Searching")
+    }
```
if `use_bookmarks` is `false` there's no need to calculate these (specifically the authorization which can be a bit slow), just adding a `if use_bookmarks` at the end of the line

Ron-Lavi replied:
Updated it :)

tbrisker commented on app/helpers/search_bar_helper.rb:
```diff
@@ -0,0 +1,25 @@
+module SearchBarHelper
+  def mount_search_bar(
+    id:,
```
should the id actually be a kwarg? i think since its required and usually the only one used you can make it a regular argument (you can mix regular arguments with kwargs, as long as the regular ones are first)

Ron-Lavi replied:
Done :)

tbrisker commented on app/views/common/_searchbar.html.erb:
```diff
@@ -1,21 +1,2 @@
-  documentationUrl: documentation_url("4.1.5Searching")
-}.to_json) %>
+<div id="search-bar"></div>
+<%= mount_search_bar(id: "search-bar"); %>
```
missing newline at end of file

Ron-Lavi replied:
Oops, wonder why hound didn't catch this error.. fixed.

tbrisker reviewed (dismissed):
just a few minor comments regarding the ruby code, didn't review react code.

ohadlevy commented on app/helpers/search_bar_helper.rb:
```diff
@@ -0,0 +1,28 @@
+        documentationUrl: documentation_url("4.1.5Searching")
+      }
+    end
+    mount_react_component('SearchBar', "##{id}", {
```
why double '##' ?

Ron-Lavi replied:
The `mount_react_component` expect a `query selector`.
In this case it's an id that while being passed to  `mount_search_bar`
doesn't need to specify the `#` char.

e.g.: 
`<div id="search-bar"></div>`
`<%= mount_search_bar("search-bar"); %>`

ohadlevy reviewed (dismissed):
Looks good @Laviro thank you! atm pending on packaging /cc @ekohl

ohadlevy commented:
@Laviro one small issue I've discovered (that can be fixed as a follow up PR) is for search query that contains a ```.```` in foreman today it should continue the auto completer vs offer the operators on mouse click, e.g.:
![recorded](https://user-images.githubusercontent.com/20167/45596354-6c356500-b9c3-11e8-8ded-22875c652d83.gif)

Ron-Lavi commented:
@ohadlevy Thanks for catching this bug.
I fix it by not adding a space when the selected menu item ends with point (".").
Do you know of any other scenarios like this one? thanks :)

ohadlevy commented:
@Laviro no, afaik thats the only special char. thanks!

ohadlevy commented:
@Laviro can you squash all of your commits? I believe we are ready to merge, simply pending on packaging ack.

Ron-Lavi commented:
@ekohl Can you check packaging ? thanks

Laviro commented:
[test foreman]

Laviro commented:
[test katello]

ohadlevy commented:
@Laviro can you please rebase? thanks

Laviro commented:
[test foreman]

Laviro commented:
[test katello]

Laviro commented:
[test foreman]

Laviro commented:
[foreman test]

Laviro commented:
[test foreman]

ohadlevy commented on package.json:
```diff
@@ -79,7 +80,7 @@
     "multiselect": "~0.9.12",
     "patternfly": "^3.42.0",
-    "patternfly-react": "2.5.1",
+    "patternfly-react": "^2.19.1",
```
@ekohl ack? this should match https://github.com/theforeman/foreman-packaging/pull/3063

mmoll commented:
test failures related?

Ron-Lavi commented:
@mmoll hey, there is one test failing.. I am about to fix it soon.

Ron-Lavi commented:
@mmol fixed the issue in `config_report`:

from:
```test "reports for host" do	
    visit config_reports_path	
    click_link(@report.host.fqdn)	
    assert_equal "host = #{@report.host.fqdn}", find_field('search').value	
  end
```

to:
```test "reports for host" do
    visit config_reports_path
    click_link(@report.host.name)
    has_selector?(".rbt-input-main", text: "host = #{@report.host.name}", wait: 3)
    assert_equal "host = #{@report.host.name}", find('.rbt-input-main').value
  end
```

Can someone review? https://github.com/theforeman/foreman/pull/5771/files#diff-5397871de10ea12805c45b89e79dfc1f

ohadlevy commented:
Ack pending tests - thanks.

Ron-Lavi commented:
All tests have passed :)

boaz0 commented:
![yesss](https://media.giphy.com/media/l41m74ATRsVYOUI92/giphy.gif)

lizagilman commented:
:tada: :tada: :tada:

Ron-Lavi commented:
Good job everyone !!!

iNecas commented on app/helpers/search_bar_helper.rb:
```diff
@@ -0,0 +1,28 @@
+  def mount_search_bar(
+    id,
+    controller: auto_complete_controller_name,
+    url: "#{auto_complete_controller_name}/auto_complete_search",
```
FYI this broke the auto_complete feature for nested controllers (that we have in some plugins). The PR to fix it is here https://github.com/theforeman/foreman/pull/6460
